src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -434,15 +434,34 @@ public void EnsureSolutionIsLoaded()
             });
         }
 
-        public string SolutionDirectory => NuGetUIThreadHelper.JoinableTaskFactory.Run(GetSolutionDirectoryAsync);
+        private string _solutionDir;
+        public string SolutionDirectory
+        {
+            get
+            {
+                if (!string.IsNullOrEmpty(_solutionDir))
```

Comment:
I see you're relying on a null, while Andy was doing the is initialized approach. 

Maybe this is better than that. 

The one scenario that Andy's implementation had issues with was the temporary project one. 

You can see which test was failing and see why.

Reply:
My PR is passing all the tests. Is it still concern?

Reply:
IMO yes, as the approaches are different. 

This particular change has undergone multiple iterations, and we've slightly been getting better. 

Maybe this is the right approach, but I'm not 100% confident I trust the tests in all scenarios.

In case of a null this retries, but Andy's didn't. The difference between 2 approaches is minor but understanding why exactly there's a difference would allow to be confident about the change.

Reply:
I'll look into Andy's PR, I'm not familiar with all the changes in that PR. 
Could we ask vendor to test this for us with different scenarios?

Reply:
Please take a look at that PRs. 

The differences between the 2 are easy to count. 

IMO, having high confidence in the changes requires us to understand why the 2 slightly different implementations are behaving differently. 

I don't believe the vendors are the going to help us testing here. 
The problems we've had are timing and more often theoretical than fully deterministic.

Reply:
Do you know where is this [Get-Package](https://github.com/NuGet/NuGet.Client/blob/84d99ce064a5f70628c89af9f30693581e84dc37/test/EndToEnd/tests/GetPackageTest.ps1#L393) defined for End2nd test? 
Not sure what is this GetPackageDoesNotThrowIfSolutionIsTemporary for now.

Reply:
Is it NuGet package manager console command?

Reply:
If you search for GetPackageCommand.cs you get https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.PowerShellCmdlets/Cmdlets/GetPackageCommand.cs.

Reply:
> It's on UI thread and initialized still throw exception, tbh I'm not familiar with VS Com interface.

It's just a native API that's being called, so the error communication is not the same. 

When `OnAfterOpenSolution` is called, is what's the value of fNewSolution https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivssolutionevents.onafteropensolution?view=visualstudiosdk-2022#parameters? 

Are the params any different from usual? 

What Andy was trying to ensure is that we don't have any synchronization problems.
Basically there are solution open/close and that's the ideal time to update the solution directory. 

I don't know if that's a real concern as long as we fallback if the value is null in the getter for Solution Directory which was added later.

Reply:
> I don't believe the vendors are the going to help us testing here. The problems we've had are timing and more often theoretical than fully deterministic.

https://github.com/NuGet/NuGet.Client/pull/4879 has more changes regarding if solution is open or not besides `solution dir`. Here in this PR we're only interested in `solution dir` value caching, change is much less.  I believe https://github.com/NuGet/Home/issues/11090 is much larger issue than this. Could we scope into current issue instead of solving both?

Reply:
> https://github.com/NuGet/NuGet.Client/pull/4879 has more changes regarding if solution is open or not besides solution dir

I'm not sure I agree. The changes are very, very similar. 

- The other PR does the initial calculation in the InitializeAsync part, https://github.com/NuGet/NuGet.Client/pull/4879/files#diff-bed53a8b2e737e49ff6141c4279e525dfa2d100eabf6510ef6e0a1746de10505R208-R216. This PR is skipping that. Think of it that this PR is retrieving the directory in a lazy way, while the other change is just doing it at the first point where the solution directory is available. 
- IsSolutionOpen check is the extra code, but it's really the same solution. If you look at the implementation it's just merely checking that the solution directory is not null. 
- The other PRs solution directory check is again relying on a cookie being set, which means InitializeAsync is being run, while you check for null value. 
- OnAfterOpenSolution is calculating the solution directory at the first time it should be available. Think of it InitializeAsync, but for new solutions potentially. 

> I believe https://github.com/NuGet/Home/issues/11090 is much larger issue than this current issue I'm working on

The solution directory calculation is equivalent in both approaches. 

I think this is change is going to work, but understanding the difference between the 2 might allow us to potentially have an even better solution. 

GetSolutionDirectory(Async) could just literally return the value. There should never been a need to do a calculation, as long as get the correct value at initialization time.

Reply:
I believe I fixed my other PR: https://github.com/NuGet/NuGet.Client/pull/4879/commits/fd029a7386b50328354b64c2e9d603712aa542ed The test passes locally, and I'm now waiting for CI results.

Therefore, I believe we can close this PR and accept mine: https://github.com/NuGet/NuGet.Client/pull/4879